### ISSUE TITLE
How to verify block header timestamp is valid?

### ISSUE THREAD

**Issue by zcc0721** [EOSIO/eos#7447]
Hi, I'm a fresh to learn eos. I want to know how to verify a block header timestamp is valid?

I read the code, and found some code:

https://github.com/EOSIO/eos/blob/cf1a2faf9b4942d6c715020120505c699da9dcab/plugins/producer_plugin/producer_plugin.cpp

```cpp
void on_incoming_block(const signed_block_ptr& block) {
         auto id = block->id();

         fc_dlog(_log, "received incoming block ${id}", ("id", id));

         EOS_ASSERT( block->timestamp < (fc::time_point::now() + fc::seconds( 7 )), block_from_the_future,
                     "received a block from the future, ignoring it: ${id}", ("id", id) );

         chain::controller& chain = chain_plug->chain();

         /* de-dupe here... no point in aborting block if we already know the block */
         auto existing = chain.fetch_block_by_id( id );
         if( existing ) { return; }

         // start processing of block
         auto bsf = chain.create_block_state_future( block );

         // abort the pending block
         chain.abort_block();

         // exceptions throw out, make sure we restart our loop
         auto ensure = fc::make_scoped_exit([this](){
            schedule_production_loop();
         });

         // push the new block
         bool except = false;
         try {
            chain.push_block( bsf );
         } catch ( const guard_exception& e ) {
            chain_plug->handle_guard_exception(e);
            return;
         } catch( const fc::exception& e ) {
            elog((e.to_detail_string()));
            except = true;
         } catch ( boost::interprocess::bad_alloc& ) {
            chain_plugin::handle_db_exhaustion();
            return;
         }

         if( except ) {
            app().get_channel<channels::rejected_block>().publish( priority::medium, block );
            return;
         }

         if( chain.head_block_state()->header.timestamp.next().to_time_point() >= fc::time_point::now() ) {
            _production_enabled = true;
         }


         if( fc::time_point::now() - block->timestamp < fc::minutes(5) || (block->block_num() % 1000 == 0) ) {
            ilog("Received block ${id}... #${n} @ ${t} signed by ${p} [trxs: ${count}, lib: ${lib}, conf: ${confs}, latency: ${latency} ms]",
                 ("p",block->producer)("id",fc::variant(block->id()).as_string().substr(8,16))
                 ("n",block_header::num_from_id(block->id()))("t",block->timestamp)
                 ("count",block->transactions.size())("lib",chain.last_irreversible_block_num())("confs", block->confirmed)("latency", (fc::time_point::now() - block->timestamp).count()/1000 ) );
         }
      }
```

It seems that valid block header timestamp < now + 7 seconds. 

but in https://github.com/EOSIO/eos/blob/cf1a2faf9b4942d6c715020120505c699da9dcab/plugins/producer_plugin/producer_plugin.cpp

```cpp
optional<fc::time_point> producer_plugin_impl::calculate_next_block_time(const account_name& producer_name, const block_timestamp_type& current_block_time) const {
   chain::controller& chain = chain_plug->chain();
   const auto& hbs = chain.head_block_state();
   const auto& active_schedule = hbs->active_schedule.producers;

   // determine if this producer is in the active schedule and if so, where
   auto itr = std::find_if(active_schedule.begin(), active_schedule.end(), [&](const auto& asp){ return asp.producer_name == producer_name; });
   if (itr == active_schedule.end()) {
      // this producer is not in the active producer set
      return optional<fc::time_point>();
   }

   size_t producer_index = itr - active_schedule.begin();
   uint32_t minimum_offset = 1; // must at least be the "next" block

   // account for a watermark in the future which is disqualifying this producer for now
   // this is conservative assuming no blocks are dropped.  If blocks are dropped the watermark will
   // disqualify this producer for longer but it is assumed they will wake up, determine that they
   // are disqualified for longer due to skipped blocks and re-caculate their next block with better
   // information then
   auto current_watermark_itr = _producer_watermarks.find(producer_name);
   if (current_watermark_itr != _producer_watermarks.end()) {
      auto watermark = current_watermark_itr->second;
      auto block_num = chain.head_block_state()->block_num;
      if (chain.pending_block_state()) {
         ++block_num;
      }
      if (watermark > block_num) {
         // if I have a watermark then I need to wait until after that watermark
         minimum_offset = watermark - block_num + 1;
      }
   }

   // this producers next opportuity to produce is the next time its slot arrives after or at the calculated minimum
   uint32_t minimum_slot = current_block_time.slot + minimum_offset;
   size_t minimum_slot_producer_index = (minimum_slot % (active_schedule.size() * config::producer_repetitions)) / config::producer_repetitions;
   if ( producer_index == minimum_slot_producer_index ) {
      // this is the producer for the minimum slot, go with that
      return block_timestamp_type(minimum_slot).to_time_point();
   } else {
      // calculate how many rounds are between the minimum producer and the producer in question
      size_t producer_distance = producer_index - minimum_slot_producer_index;
      // check for unsigned underflow
      if (producer_distance > producer_index) {
         producer_distance += active_schedule.size();
      }

      // align the minimum slot to the first of its set of reps
      uint32_t first_minimum_producer_slot = minimum_slot - (minimum_slot % config::producer_repetitions);

      // offset the aligned minimum to the *earliest* next set of slots for this producer
      uint32_t next_block_slot = first_minimum_producer_slot  + (producer_distance * config::producer_repetitions);
      return block_timestamp_type(next_block_slot).to_time_point();
   }
}
```

https://github.com/EOSIO/eos/blob/cf1a2faf9b4942d6c715020120505c699da9dcab/plugins/producer_plugin/producer_plugin.cpp

```cpp
fc::time_point producer_plugin_impl::calculate_pending_block_time() const {
   const chain::controller& chain = chain_plug->chain();
   const fc::time_point now = fc::time_point::now();
   const fc::time_point base = std::max<fc::time_point>(now, chain.head_block_time());
   const int64_t min_time_to_next_block = (config::block_interval_us) - (base.time_since_epoch().count() % (config::block_interval_us) );
   fc::time_point block_time = base + fc::microseconds(min_time_to_next_block);


   if((block_time - now) < fc::microseconds(config::block_interval_us/10) ) {     // we must sleep for at least 50ms
      block_time += fc::microseconds(config::block_interval_us);
   }
   return block_time;
}
```

so many functions to calculate valid timestamp, I just want to know which way is right way to verify block header timestamp?

Please help me!! Thank you very much.

**Comment by zcc0721** [comment]
Can any one show the right code block to solve this problem? Thanks a lot.

**Comment by zcc0721** [comment]
I want to know why `block->timestamp < (fc::time_point::now() + fc::seconds( 7 ))` , and one producer create  12 new blocks in 6 seconds, why this is 7 seconds? 

Thank you.

**Comment by aaron-fu-block-one** [comment]
Hi @zcc0721.

In general, the consensus on whether a block is valid or not is reached via the DPOS algorithm on EOS. See [here](https://steemit.com/dpos/@dantheman/dpos-consensus-algorithm-this-missing-white-paper) for a detailed introduction.

Regarding the timestamps in the block header, there can be additional checks that serve as sanity checks, in order to rule out the obviously false information that goes certainly against the logic. For example,

- existing blocks must come from the past
- next block must come from the future
- next block cannot come too far away from the future

They are necessary but not sufficient conditions to ensure a block is valid.

------

Regarding the code you quoted

```c++
block->timestamp < (fc::time_point::now() + fc::seconds( 7 )), 
block_from_the_future,
"received a block from the future, ignoring it: ${id}", ("id", id)
```

from the `on_incoming_block()` function in the `producer_plugin.cpp`, it is called when (and only when) the producer plugin is initialized.

Note that `fc::time_point::now()` refers to the time on the local machine, which could be inaccurate to various degrees, and therefore some error tolerance should be given.

**On a side note:** That `block_from_the_future` expection is derived from `block_validate_exception`, which derives from `chain_exception`. If you look at the `exceptions.hpp` in the `libraries/chain/include/eosio/chain/`, you will find another exception `block_too_old_exception` which could complain just the opposite way: *block is too old to push*.

------

There are other cases where timestamps are checked.

For example, in the `\libraries\chain\block_header_state.cpp`, you may find

```c++
pending_block_header_state
block_header_state::next( block_timestamp_type when,
                          uint16_t num_prev_blocks_to_confirm )const
{
   pending_block_header_state result;
   if( when != block_timestamp_type() ) {
      EOS_ASSERT( when > header.timestamp, block_validate_exception,
                  "next block must be in the future" );
        ...
```

In the same file, there is another timestamp check, for consistency.

```c++
block_header_state pending_block_header_state::_finish_next(
                                 const signed_block_header& h,
                                 const std::function<void( block_timestamp_type,
                                 const flat_set<digest_type>&,
                                 const vector<digest_type>& )>& validator )&&
{
   EOS_ASSERT( h.timestamp == timestamp,
                                 block_validate_exception,
                                "timestamp mismatch" );
   ...
```

To sum up, validation for blocks is primarily achieved via the DPOS algorithm, whereas the timestamp checks take place when needed as additional sanity checks.

------

Regarding why it is 7 seconds for `block->timestamp < (fc::time_point::now() + fc::seconds( 7 )) ...`, it could be a subjective configuration by the code designer. I will come back to you if I have more details on this. Thank you.

See also for related discussions on Ethereum:

https://ethereum.stackexchange.com/questions/45026/check-that-the-timestamp-of-the-block-is-greater-than-that-of-the-previous-block

**Comment by b1aafulei** [comment]
I am closing this issue. If anyone has a similar question and would like to have more discussion, please feel free to reopen it.
